src/Build/BackEnd/Components/RequestBuilder/AssemblyLoadsTracker.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+
+using System;
+using System.Collections.Concurrent;
+using Microsoft.Build.BackEnd.Logging;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.Build.BackEnd.Components.RequestBuilder
+{
+    internal class AssemblyLoadsTracker : MarshalByRefObject, IDisposable
+    {
+        private static readonly ConcurrentDictionary<AppDomain, AssemblyLoadsTracker> s_instances = new();
+        private readonly LoggingContext? _loggingContext;
+        private readonly LoggingService? _loggingService;
+        private readonly AssemblyLoadingContext _context;
+        private readonly AppDomain _appDomain;
+
+        private AssemblyLoadsTracker(
+            LoggingContext? loggingContext,
+            LoggingService? loggingService,
+            AssemblyLoadingContext context,
+            AppDomain appDomain)
+        {
+            _loggingContext = loggingContext;
+            _loggingService = loggingService;
+            _context = context;
+            _appDomain = appDomain;
+        }
+
+        public static IDisposable StartTracking(
+            LoggingContext loggingContext,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
+            return StartTracking(loggingContext, null, context, appDomain);
+        }
+
+        public static IDisposable StartTracking(
+            LoggingService loggingService,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
+            return StartTracking(null, loggingService, context, appDomain);
+        }
+
+        public static void StopTracking(AppDomain appDomain)
+        {
+            if (s_instances.TryRemove(appDomain, out AssemblyLoadsTracker? tracker))
+            {
+                tracker.StopTracking();
+            }
+        }
+
+        public void Dispose()
+        {
+            StopTracking();
+        }
+
+        private static IDisposable StartTracking(
+            LoggingContext? loggingContext,
+            LoggingService? loggingService,
+            AssemblyLoadingContext context,
+            AppDomain? appDomain = null)
+        {
+            var tracker = new AssemblyLoadsTracker(loggingContext, loggingService, context, appDomain ?? AppDomain.CurrentDomain);
+            if (appDomain != null)
+            {
+                s_instances.AddOrUpdate(appDomain, tracker, (_, loadsTracker) => loadsTracker);
```

Comment:
The `AddOrUpdate` call seems suspicious here. In the case there is already an existing entry the value `tracker` is essentially lost. It will get a call to `StartTracking` but never `StopTracking`. I don't know the significance of that in the MSBuild system but it seems better to avoid if we can. 

It feels like the code here should be doing one of the following: 

1. Not calling `StartTracking` then the`AssemblyLoadsTracker` instance is not in the list
2. Having `s_instances` store a list vs. a single value